### PR TITLE
Do not offer `string` (or other type-keywords) after `ref/readonly` inside namespace member declaration

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/67061

This is both not legal, and interferes with typing `ref struct` since `string` sorts above `struct`.

### CODE REVIEW

**Comment by DoctorKrolic** [comment]
Wait a minute, why is `string` special here? I thought, such logic can be applied to all type names, `string` is just the most noticeable one due to interfering with `struct`, isn't it?

**Comment by CyrusNajmabadi** [comment]
> Wait a minute, why is string special here? I thought, such logic can be applied to all type names, string is just the most noticeable one due to interfering with struct, isn't it?

Because the others don't interfere with typing.  So i'm only addressing the core case that is problematic.

**Comment by DoctorKrolic** [comment]
Maybe pull that logic up to `AbstractSpecialTypePreselectingKeywordRecommender` to achieve consistency at least among special-type keywords?

**Comment by CyrusNajmabadi** [comment]
Sure.  i can try that.

**Comment by CyrusNajmabadi** [comment]
Moved up.  Thanks :)

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSpecialTypePreselectingKeywordRecommender.cs [hunk 1]
```diff
@@ -36,7 +38,42 @@ protected sealed override bool IsValidContext(int position, CSharpSyntaxContext
...
+            }
+
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            token = token.GetPreviousTokenIfTouchingWord(position);
+
+            // if we have `readonly` move backwards to see if we have `ref readonly`.
+            if (token.Kind() is SyntaxKind.ReadOnlyKeyword)
+                token = syntaxTree.FindTokenOnLeftOfPosition(token.SpanStart, cancellationToken);
+
+            // if we're not after `ref` or `ref readonly` then don't offer `string` here.
```
Comment (DoctorKrolic) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSpecialTypePreselectingKeywordRecommender.cs:61]:
```suggestion
            // if we're not after `ref` or `ref readonly` then don't offer keyword here.
```

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSpecialTypePreselectingKeywordRecommender.cs [hunk 2]
```diff
@@ -36,7 +38,42 @@ protected sealed override bool IsValidContext(int position, CSharpSyntaxContext
...
+            if (token.Kind() != SyntaxKind.RefKeyword)
+                return false;
+
+            // If we're inside a type, this is always to have a ref/readonly string.
```
Comment (DoctorKrolic) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSpecialTypePreselectingKeywordRecommender.cs:65]:
```suggestion
            // If we're inside a type, this is always to have a ref/readonly type name.
```

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSpecialTypePreselectingKeywordRecommender.cs [hunk 3]
```diff
@@ -36,7 +38,42 @@ protected sealed override bool IsValidContext(int position, CSharpSyntaxContext
...
+            var containingType = token.GetAncestor<TypeDeclarationSyntax>();
+            if (containingType != null)
+                return true;
+
+            // If not in a type, but in a namespace, this is not ok to have a ref/readonly string.
```
Comment (DoctorKrolic) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSpecialTypePreselectingKeywordRecommender.cs:70]:
```suggestion
            // If not in a type, but in a namespace, this is not ok to have a ref/readonly type name.
```

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSpecialTypePreselectingKeywordRecommender.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSpecialTypePreselectingKeywordRecommender.cs:61]:
```suggestion
            // if we're not after `ref` or `ref readonly` then don't offer a type-keyword here.
```

**Review by akhera99 — APPROVED** [review]

**Pull request merged**